### ISSUE TITLE
Add no-unnecessary-type-check rule

### ISSUE THREAD

**Issue by zachkirsch** [palantir/tslint#4733]
### Rule Suggestion

**Is your rule for a general problem or is it specific to your development style?**

General problem

**What does your suggested rule do?**

Warns you if you compare against `null` or `undefined` unnecessarily.

**List several examples where your rule could be used**

```typescript
const x = "my string"; // definitely not undefined or null

if (x !== undefined) { // should warn
    console.log(x);
}
if (x !== null) { // should warn
    console.log(x);
}

function getStringOrUndefined() {
    return Math.random() > 0.5 ? "string" : undefined;
}

const y: string | undefined = getStringOrUndefined();
if (y !== undefined) { // should not warn
    console.log(y);
}
```

**Additional context**

Checking if a variable is `undefined` unnecessarily is a problem for two reasons. Either the variable can actually never be `undefined` - in which case the check is unnecessary and probably confusing to a reader. Or, the variable _can_ be undefined (i.e. the code-writer knows something that the compiler doesn't) - in which case the typings should be updated to reflect that the variable _might_ be undefined.

In the former case, this rule more more for style. But in the latter (and imo more important) case, this rule can prevent bugs - i.e. since the type is wrong, there could be a variable of that type somewhere that could be `undefined` but neither the person using the variable nor the compiler know this.

Submitting a PR with this rule shortly

**Comment by JoshuaKGoldberg** [comment]
🤔 this seems like a subset of `strict-type-predicates`: https://palantir.github.io/tslint/rules/strict-type-predicates/

Tempting to close as a duplicate and already resolved... Thoughts @zachkirsch?

**Comment by zachkirsch** [comment]
Oh man! I guess i should have made the issue first..

Well it was a good learning experience :)
